mux.js

```diff
@@ -122,90 +134,112 @@ Multiplexer.prototype._finalize = function(err) {
+  var filtered = keys.filter(function(key) {
+    if (self._localOffer.indexOf(key) === -1) {
+      debug('[REPLICATION] Warning, remote requested feed that is not in offer', key)
+      return false
+    }
 
-  var sending = self._remoteWants.filter(function(k) {
-    return (self._localHave || []).indexOf(k) !== -1
+    // All good, we accept the key request
+    return true
   })
+  filtered = uniq(filtered)
+  // Tell remote which keys we will replicate
+  debug('[REPLICATION] Sending REPLICICATE_FEEDS')
+  this._feed.extension(REPLICATE_FEEDS, Buffer.from(JSON.stringify(filtered)))
 
-  var receiving = self._localWant.filter(function(k){
-    return (self._remoteHas || []).indexOf(k) !== -1
-  })
+  // Start replicating as promised.
+  this._replicateFeeds(filtered)
+}
 
-  // Concat sending and receiveing; produce sorted array with no duplicates
-  var keys = sending.concat(receiving)
-    .reduce(function(arr, key){ // remove duplicates
-      if (arr.indexOf(key) === -1) arr.push(key)
-      return arr
-    }, [])
-    .sort() // sort
+Multiplexer.prototype._onRemoteReplicate = function (keys) {
+  var self = this
+  var filtered = keys.filter(function(key) {
+    return self._requestedFeeds.indexOf(key) !== -1
+  })
 
-  debug(this._id + ' [REPLICATION] _initRepl', keys.length, keys)
+  // Start replicating as requested.
+  this._replicateFeeds(filtered)
+}
 
-  // End immedietly if there's nothing to replicate.
-  if (!this._opts.live && keys.length === 0) return this._finalize()
+// Initializes new replication streams for feeds and joins their streams into
+// the main stream.
+Multiplexer.prototype._replicateFeeds = function(keys) {
+  var self = this
+  keys = uniq(keys)
+  debug(this._id + '[REPLICATION] _replicateFeeds', keys.length, keys)
 
   this.emit('replicate', keys, startFeedReplication)
 
   return keys
 
   function startFeedReplication(feeds){
     if (!Array.isArray(feeds)) feeds = [feeds]
-    self.stream.expectedFeeds = feeds.length
+    self.stream.expectedFeeds += feeds.length
 
     // only the feeds passed to `feeds` option will be replicated (sent or received)
     // hypercore-protocol has built in protection against receiving unexpected/not asked for data.
     feeds.forEach(function(feed) {
-      feed.ready(function() { // wait for each to be ready before replicating.
-        debug(self._id + ' [REPLICATION] replicating feed:', feed.key.toString('hex'))
-        feed.replicate(Object.assign({}, {
+      feed.ready(function() { // wait for each feed to be ready before replicating.
+        var hexKey = feed.key.toString('hex');
+
+        // prevent a feed from being folded into the main stream twice.
+        if (typeof self._activeFeedStreams[hexKey] !== 'undefined') {
+          debug(self._id + '[REPLICATION] warning! Prevented duplicate replication of: ', hexKey)
+          // decrease the expectedFeeds that was unconditionally increased
+          self.stream.expectedFeeds.length--
```

Comment:
I'm confused. Is `expectedFeeds` supposed to be a number or an array? I see `.length` decremented here, but `expectedFeeds` itself incremented above. More troubling: no tests are failing. O___O